diff --git a/localedef/include/bits/floatn-common.h b/localedef/include/bits/floatn-common.h
new file mode 100644
index 0000000..7c1c550
--- /dev/null
+++ b/localedef/include/bits/floatn-common.h
@@ -0,0 +1 @@
+#include "glibc/bits/floatn-common.h"
diff --git a/localedef/include/bits/floatn.h b/localedef/include/bits/floatn.h
new file mode 100644
index 0000000..da4c9e3
--- /dev/null
+++ b/localedef/include/bits/floatn.h
@@ -0,0 +1 @@
+#include "glibc/bits/floatn.h"
diff --git a/localedef/include/bits/long-double.h b/localedef/include/bits/long-double.h
new file mode 100644
index 0000000..6092662
--- /dev/null
+++ b/localedef/include/bits/long-double.h
@@ -0,0 +1 @@
+#include "glibc/bits/long-double.h"
