diff --git a/Makefile b/Makefile
index 2a074ce..ea0af99 100644
--- a/Makefile
+++ b/Makefile
@@ -8,7 +8,7 @@
 #LD=${CROSSCOMPILE}ld
 #OBJCOPY=${CROSSCOMPILE}objcopy
 #OBJDUMP=${CROSSCOMPILE}objdump
-CFLAGS=-I. -O2 -ggdb -march=rv64imafdc -mabi=lp64d -Wall -mcmodel=medany -mexplicit-relocs -Wno-main
+CFLAGS=-I. -O2 -ggdb -march=rv64imafdc -mabi=lp64d -Wall -mcmodel=medany -mexplicit-relocs -Wno-main -no-pie -fno-PIE
 CCASFLAGS=-I. -mcmodel=medany -mexplicit-relocs
 LDFLAGS=-nostdlib -nostartfiles
 
