diff --git a/Makefile b/Makefile
index ea0af99..b364187 100644
--- a/Makefile
+++ b/Makefile
@@ -8,7 +8,7 @@
 #LD=${CROSSCOMPILE}ld
 #OBJCOPY=${CROSSCOMPILE}objcopy
 #OBJDUMP=${CROSSCOMPILE}objdump
-CFLAGS=-I. -O2 -ggdb -march=rv64imafdc -mabi=lp64d -Wall -mcmodel=medany -mexplicit-relocs -Wno-main -no-pie -fno-PIE
+CFLAGS=-I. -O2 -ggdb -march=rv64imafdc -mabi=lp64d -Wall -mcmodel=medany -mexplicit-relocs -Wno-main -no-pie -fno-PIE -fno-tree-loop-distribute-patterns
 CCASFLAGS=-I. -mcmodel=medany -mexplicit-relocs
 LDFLAGS=-nostdlib -nostartfiles
 
