diff --git a/Makefile b/Makefile
index d2fa53a..83dcf72 100644
--- a/Makefile
+++ b/Makefile
@@ -3,11 +3,11 @@
 # SPDX-License-Identifier: GPL-2.0-or-later
 # See the file LICENSE for further information
 
-CROSSCOMPILE?=riscv64-unknown-elf-
-CC=${CROSSCOMPILE}gcc
-LD=${CROSSCOMPILE}ld
-OBJCOPY=${CROSSCOMPILE}objcopy
-OBJDUMP=${CROSSCOMPILE}objdump
+#CROSSCOMPILE?=riscv64-unknown-elf-
+#CC=${CROSSCOMPILE}gcc
+#LD=${CROSSCOMPILE}ld
+#OBJCOPY=${CROSSCOMPILE}objcopy
+#OBJDUMP=${CROSSCOMPILE}objdump
 CFLAGS=-I. -O2 -ggdb -march=rv64imafdc -mabi=lp64d -Wall -mcmodel=medany -mexplicit-relocs -Wno-main
 CCASFLAGS=-I. -mcmodel=medany -mexplicit-relocs
 LDFLAGS=-nostdlib -nostartfiles
